File: 1. Two Sum.py
#PYTHON
#* 1. Two Sum
nums = [2,7,11,15]; target = 9
from typing import List
#* MY SOL 1
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]
                
print(Solution().twoSum(nums, target)) #[0, 1]
                
#* MY SOL 1.1 | ONE LINER
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        return next(([i, j] for i in range(len(nums)) for j in range(i + 1, len(nums)) if nums[i] + nums[j] == target))
                                      #== target), None) <= if u need else

print(Solution().twoSum(nums, target))

#* ENUMERATE WITH_INDEX FOR SHOW/LEARNING | NOT PROPER
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i, num in enumerate(nums):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]
                
print(Solution().twoSum(nums, target))

File: 1. Two Sum.rs
// 1. Two Sum
let nums = vec![2, 7, 11, 15];
let target = 9;
//FAST
use std::collections::HashMap;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        
        for (i, &num) in nums.iter().enumerate() {
            if let Some(&j) = map.get(&(target - num)) {
                return vec![i as i32, j as i32];
            }
            map.insert(num, i);
        }
        
        vec![]
    }
}
let result_fast = Solution::two_sum(nums.clone(), target);
println!("FAST: {:?}", result_fast);

//ORIG
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
      for i in 0..nums.len() {
        for j in (i+1)..nums.len() {
          if nums[i] + nums[j] == target{
            return vec![i as i32,j as i32]  
            } 
        }
      }
        return vec![]
    }
}
let result = Solution::two_sum(nums.clone(), target);
println!("{:?}", result);